File: ai/prompts.py
"""Prompt templates for AI analysis."""


def text_analysis_prompt(text, language='English'):
    return f"""Analyze this {language} text from a student assessment for spelling errors, homophone misuse, and grammar issues.

Text:
{text}

Return ONLY a JSON object with two fields:
- "summary": A brief description of what was analyzed and what was found (1-2 sentences).
- "issues": A JSON array of issues. Each item: {{"type": "spelling|homophone|grammar", "text": "wrong word", "suggestion": "correct word", "context": "surrounding text"}}

If no issues, return: {{"summary": "Analyzed [brief description]. No issues found.", "issues": []}}
Example: {{"summary": "Analyzed 3 paragraphs of English assessment text. Found 2 spelling errors.", "issues": [...]}}"""


def vision_analysis_prompt(context=''):
    ctx_part = f"\nContext: {context}" if context else ""
    return f"""Analyze this screenshot from a student assessment for visual quality issues.{ctx_part}

Check for: text readability, layout problems, missing content, contrast issues, rendering artifacts.

Return ONLY a JSON object with two fields:
- "summary": A brief description of what was checked and what was found (1-2 sentences).
- "issues": A JSON array of issues. Each item: {{"type": "readability|layout|missing|contrast|artifact", "detail": "description", "severity": "low|medium|high"}}

If no issues, return: {{"summary": "Checked [brief description]. No visual issues detected.", "issues": []}}
Example: {{"summary": "Checked assessment item screenshot for visual quality. Found 1 contrast issue in the header.", "issues": [...]}}"""


def text_comparison_prompt(baseline, current, language='English'):
    return f"""Compare these two versions of {language} assessment text and identify significant differences.

Baseline:
{baseline}

Current:
{current}

Return ONLY a JSON array of differences. Each item: {{"type": "added|removed|changed|reordered", "baseline": "...", "current": "...", "significance": "low|medium|high"}}
If no differences: []"""


RESPONSE_FORMAT_INSTRUCTIONS = """

Return ONLY a JSON object with two fields:
- "summary": A brief description of what was checked and what was found, including specifics about any issues (1-2 sentences).
- "issues": A JSON array of issues. Each item: {"type": "category", "detail": "description of the issue", "severity": "low|medium|high"}

If no issues, return: {"summary": "Checked [brief description]. No issues detected.", "issues": []}
Example: {"summary": "Analyzed text. Found 2 issues.", "issues": [{"type": "spelling", "detail": "...", "severity": "low"}]}"""


def wrap_custom_prompt(custom_prompt, context=''):
    """Wrap a user-defined custom prompt with standard JSON response format instructions."""
    ctx_part = f"\nContext: {context}" if context else ""
    return f"{custom_prompt}{ctx_part}{RESPONSE_FORMAT_INSTRUCTIONS}"


def test_generation_system_prompt(helpers=None):
    helper_section = ''
    if helpers:
        helper_lines = []
        for fname, funcs in helpers.items():
            for f in funcs:
                helper_lines.append(f"- {f['name']} (from {fname})")
        if helper_lines:
            helper_section = '\n\nAvailable helpers:\n' + '\n'.join(helper_lines)

    return f"""You are SCOUT AI, an expert at writing Playwright test scripts for the NAEP assessment platform.

Generate clean, well-commented Playwright tests using:
- @playwright/test framework
- Helper functions from src/helpers/
- test.describe() blocks with descriptive names
- Proper assertions (toBeVisible, toHaveText)
- page.screenshot() for captures (SCOUT handles baseline comparison — do NOT use toHaveScreenshot)
- @smoke, @visual, @content, @feature tags{helper_section}

Return ONLY the JavaScript code, no explanation."""

File: playwright/src/ai/prompts.js
// SCOUT — Shared Prompt Templates
// Provider-agnostic prompts used by all AI providers.

/**
 * Text analysis prompt — spelling, grammar, homophone detection.
 */
function textAnalysisPrompt(text, language = 'English') {
  return `You are a strict proofreading assistant for educational assessment items written in ${language}.

Analyze the text below for ONLY these concrete issues:
1. **Spelling errors** — misspelled words (not formatting or style)
2. **Homophone misuse** — wrong word used (e.g., "their" vs "there", "affect" vs "effect")
3. **Grammar errors** — subject-verb disagreement, broken sentence structure
4. **Garbled text** — corrupted characters, extra whitespace within words, encoding artifacts

IMPORTANT RULES:
- Only report issues you are highly confident about.
- Do NOT flag stylistic choices, mathematical notation, variable names, or assessment-specific formatting.
- Do NOT flag extra whitespace between words — only within words.
- Respond with ONLY a JSON array — no explanation, no markdown, no commentary.
- If no issues exist, respond with exactly: []

Each issue in the array must have these fields:
{"type": "spelling|homophone|grammar|garbled", "text": "the problematic text", "suggestion": "the correction", "context": "short snippet showing where it appears"}

Text to analyze:
"""
${text}
"""`;
}

/**
 * Vision analysis prompt — screenshot readability and layout check.
 */
function visionAnalysisPrompt(context = '') {
  return `You are a visual QA analyst for educational assessment items. Analyze this screenshot for ONLY significant visual defects:

1. **readability** — Blurry, overlapping, or cut-off text that prevents reading
2. **layout** — Misaligned or overlapping UI elements that break the interface
3. **contrast** — Text that is unreadable due to insufficient color contrast
4. **rendering** — Broken images, missing icons, garbled characters, visual artifacts
5. **completeness** — Visibly truncated or missing content

${context ? `Additional context: ${context}` : ''}

IMPORTANT RULES:
- Only report genuine defects that would impact a student taking this assessment.
- Do NOT flag minor cosmetic details, normal UI chrome, or intentional design choices.
- Respond with ONLY a JSON array — no explanation, no markdown, no commentary.
- If no issues exist, respond with exactly: []

Each issue: {"type": "readability|layout|contrast|rendering|completeness", "detail": "description", "severity": "high|medium|low"}`;
}

/**
 * Text comparison prompt — compare two versions of assessment text.
 */
function textComparisonPrompt(baselineText, currentText, language = 'English') {
  return `Compare these two versions of the same ${language} assessment item text and identify meaningful differences. Ignore whitespace-only changes.

**Previous version (baseline):**
"""
${baselineText}
"""

**Current version:**
"""
${currentText}
"""

Return your findings as a JSON array. Each difference should have:
- "type": one of "added", "removed", "changed", "reworded"
- "baseline": the original text (if applicable)
- "current": the new text (if applicable)
- "significance": "high" (meaning change), "medium" (minor rewording), "low" (cosmetic)

If the texts are identical or only differ by whitespace, return exactly: []`;
}

/**
 * Test generation system prompt — used by the AI test builder.
 * @param {object} helperDocs - Object mapping helper names to their descriptions
 */
function testGenerationSystemPrompt(helperDocs = {}) {
  const helperList = Object.entries(helperDocs)
    .map(([name, desc]) => `- ${name} — ${desc}`)
    .join('\n');

  return `You are a Playwright test script generator for SCOUT, the NAEP assessment testing system.

Available helpers (import from the helpers directory):
${helperList || `- loginAndNavigate(page, url) — logs in and navigates to the given URL
- getItemUrl(itemId) — returns the URL for an assessment item
- setZoom(page, percent) — sets the browser zoom level
- setTheme(page, theme) — switches theme: 'light', 'dark', 'low-contrast'
- switchLanguage(page, lang) — toggles to 'English' or 'Spanish'
- openCalculator(page) — opens the calculator overlay, returns the calculator locator
- openHelp(page) — opens the help panel
- extractItemText(page) — extracts visible text content from the item DOM`}

Conventions:
- Use require('@playwright/test') for test and expect
- Use page.screenshot() for captures — SCOUT handles baseline comparison (do NOT use toHaveScreenshot)
- Name screenshots descriptively: test-results/item-{id}-{context}.png
- Group related tests in test.describe() blocks
- Add a comment header with generation date and purpose
- Tag tests with @smoke, @visual, @content, @feature as appropriate

Generate ONLY the test file code. Use the helpers — do not rewrite login/navigation logic.`;
}

module.exports = {
  textAnalysisPrompt,
  visionAnalysisPrompt,
  textComparisonPrompt,
  testGenerationSystemPrompt,
};
